src/particle.rs

```diff
@@ -0,0 +1,431 @@
+use std::io::Write;
+
+use anyhow::Context;
+use vek::Vec3;
+
+use crate::block::{BlockPos, BlockState};
+use crate::protocol::{Decode, Encode, VarInt};
+
+#[derive(Clone, Debug)]
+pub struct Particle {
+    pub particle_type: ParticleType,
+    pub long_distance: bool,
+    pub position: Vec3<f64>,
+    pub offset: Vec3<f32>,
+    pub max_speed: f32,
+    pub particle_count: u32,
+}
+
+#[derive(Clone, Debug)]
+pub enum ParticleType {
```

Comment:
Would it be possible to add and extractor and make this enum with codegen instead?

Reply:
I was hoping this would be possible.
I don't know how the extractors work, but here is a sample of how particles are defined in MC source:
```java
        public VibrationParticleEffect read(ParticleType<VibrationParticleEffect> particleType, StringReader stringReader) throws CommandSyntaxException {
            stringReader.expect(' ');
            float f = (float)stringReader.readDouble();
            stringReader.expect(' ');
            float g = (float)stringReader.readDouble();
            stringReader.expect(' ');
            float h = (float)stringReader.readDouble();
            stringReader.expect(' ');
            int i = stringReader.readInt();
            BlockPos blockPos = new BlockPos(f, g, h);
            return new VibrationParticleEffect(new BlockPositionSource(blockPos), i);
        }
```